src/modules/push-notifications/push.notifications.service.ts

```diff
@@ -0,0 +1,240 @@
+            contractKeysRaw?.data?.pairs || {},
+        );
+
+        const userEnergyKey = BinaryUtils.stringToHex('userEnergy');
+        const userEnergyKeys = contractPairs
+            .filter(([key, _]) => key.startsWith(userEnergyKey))
+            .map(([key, _]) => key.replace(userEnergyKey, ''));
+
+        const userEnergyAddresses = userEnergyKeys.map((key) =>
+            AddressUtils.bech32Encode(key),
+        );
+
+        return userEnergyAddresses.map((address) => ({
+            address,
+            notificationSent: false,
+        }));
+    }
+
+    @ErrorLoggerAsync()
+    private async getUsersFromIndexer(
+        epoch: number,
+    ): Promise<UserEnergyAddress[]> {
+        const query = ElasticQuery.create()
+            .withPagination({ from: 0, size: 10000 })
+            .withMustExistCondition('energyDetails');
+
+        query.condition.must = [
+            QueryType.Range('energyDetails.amount', { key: 'gt', value: 0 }),
+        ];
+
+        const allAddresses: string[] = [];
+
+        await this.elasticService.getScrollableList(
+            `accounts-000001_${epoch}`,
+            'address',
+            query,
+            async (items: AccountType[]) => {
+                const addresses = items.map(
+                    (item: AccountType) => item.address,
+                );
+                allAddresses.push(...addresses);
+            },
+        );
+
+        return allAddresses.map((address) => ({
+            address,
+            notificationSent: false,
+        }));
+    }
+
+    private async sendBatchNotifications(
+        addresses: string[],
+    ): Promise<boolean> {
+        const chainId = Number(this.apiConfigService.getChainId());
+        const notificationsApiUrl =
+            this.apiConfigService.getNotificationsApiUrl();
+        const notificationsApiKey =
+            this.apiConfigService.getNotificationsApiKey();
+
+        const payload: NotificationPayload = {
+            addresses,
+            chainId,
+            title: 'Energy Update',
+            body: 'You can now claim your rewards',
+            route: '/portfolio',
+            iconUrl:
+                'https://xexchange.com/assets/imgs/mx-logos/xexchange-logo@2x.webp',
+        };
+
+        const response = await axios.post(notificationsApiUrl, payload, {
+            headers: {
+                'Content-Type': 'application/json',
+                'x-notifications-api-key': notificationsApiKey,
+            },
+        });
+
+        if (response.status !== 201) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private async processNotificationBatch(
+        addresses: string[],
+        maxRetries: number,
+    ): Promise<{ successful: string[]; failed: string[] }> {
+        const batchSize =
+            await this.apiConfigService.getNotificationsBatchSize();
+
+        const successful: string[] = [];
+        let remainingAddresses = [...addresses];
+
+        for (
```

Comment:
You can have a different approach for the retry; Use this method recursively
Have a look how doGetGeneric is implemented